db/experimental.cc

```diff
@@ -143,5 +151,1057 @@ Status UpdateManifestForFilesState(
+    append_to.push_back(kCategoryScopeFilterWrapper);
+
+    PutVarint64(&append_to, CategorySetToUint(categories_));
+
+    // Wrapping just 1 filter for now
+    PutVarint64(&append_to, 1);
+    wrapped_->Finish(append_to);
+  }
+
+ private:
+  KeySegmentsExtractor::KeyCategorySet categories_;
+  std::unique_ptr<SstQueryFilterBuilder> wrapped_;
+};
+
+class BytewiseMinMaxSstQueryFilterConfig : public SstQueryFilterConfigImpl {
+ public:
+  using SstQueryFilterConfigImpl::SstQueryFilterConfigImpl;
+
+  std::unique_ptr<SstQueryFilterBuilder> NewBuilder(
+      bool sanity_checks) const override {
+    auto b = std::make_unique<MyBuilder>(*this, sanity_checks);
+    if (categories_ != KeySegmentsExtractor::KeyCategorySet::All()) {
+      return std::make_unique<CategoryScopeFilterWrapperBuilder>(categories_,
+                                                                 std::move(b));
+    } else {
+      return b;
+    }
+  }
+
+  static bool RangeMayMatch(
+      const Slice& filter, const Slice& lower_bound_incl,
+      const KeySegmentsExtractor::Result& lower_bound_extracted,
+      const Slice& upper_bound_excl,
+      const KeySegmentsExtractor::Result& upper_bound_extracted) {
+    assert(!filter.empty() && filter[0] == kBytewiseMinMaxFilter);
+    if (filter.size() <= 4) {
+      // Missing some data
+      return true;
+    }
+    bool empty_included = (filter[1] & kEmptySeenFlag) != 0;
+    const char* p = filter.data() + 2;
+    const char* limit = filter.data() + filter.size();
+
+    FilterInput in;
+    p = DeserializeFilterInput(p, limit, &in);
+    if (p == nullptr) {
+      // Corrupt or unsupported
+      return true;
+    }
+
+    uint32_t smallest_size;
+    p = GetVarint32Ptr(p, limit, &smallest_size);
+    if (p == nullptr || static_cast<size_t>(limit - p) <= smallest_size) {
+      // Corrupt
+      return true;
+    }
+    Slice smallest = Slice(p, smallest_size);
+    p += smallest_size;
+
+    size_t largest_size = static_cast<size_t>(limit - p);
+    Slice largest = Slice(p, largest_size);
+
+    Slice lower_bound_input, lower_bound_leadup;
+    Slice upper_bound_input, upper_bound_leadup;
+    GetFilterInput(in, lower_bound_incl, lower_bound_extracted,
+                   &lower_bound_input, &lower_bound_leadup);
+    GetFilterInput(in, upper_bound_excl, upper_bound_extracted,
+                   &upper_bound_input, &upper_bound_leadup);
+
+    if (lower_bound_leadup.compare(upper_bound_leadup) != 0) {
+      // Unable to filter range when bounds have different lead-up to key
+      // segment
+      return true;
+    }
+
+    if (empty_included && lower_bound_input.empty()) {
+      // May match on 0-length segment
+      return true;
+    }
+    // TODO: potentially fix upper bound to actually be exclusive
+
+    // May match if both the upper bound and lower bound indicate there could
+    // be overlap
+    return upper_bound_input.compare(smallest) >= 0 &&
```

Comment:
Since upper bound is exclusive, if it's the same as `smallest`, can we also filter out the table?

Reply:
I've expanded the TODO to record why that change isn't safe:
```
    // TODO: potentially fix upper bound to actually be exclusive, but it's not
    // as simple as changing >= to > below, because it's upper_bound_excl that's
    // exclusive, and the upper_bound_input part extracted from it might not be.
```